### PR TITLE
Profiler: Add source code view

### PR DESCRIPTION
**Pull request by sunverwerth**
Bildschirmfoto zu 2021-12-26 23-39-25

This adds a new view mode to profiler which displays source lines and
samples that occured at those lines. This view can be opened via the
menu or by pressing CTRL-S.

It does this by mapping file names from DWARF to "/usr/src/serenity/..."
i.e. source code should be copied to /usr/src/serenity/Userland and
/usr/src/serenity/Kernel to be visible in this mode.

Currently *all* files contributing to the selected function are loaded
completely which could be a lot of data when dealing with lots of
inlined code.

### CODE REVIEW

File: Userland/DevTools/Profiler/SourceModel.cpp [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+#include "SourceModel.h"
+#include "Profile.h"
+#include <LibDebug/DebugInfo.h>
+#include <LibGUI/Painter.h>
+#include <LibSymbolication/Symbolication.h>
+#include <stdio.h>
+
+namespace Profiler {
+
+static constexpr StringView const c_source_root_path = "/usr/src/serenity/"sv;
```
Comment (Hendiadyoin1) [inline Userland/DevTools/Profiler/SourceModel.cpp:16]: This does not have to be marked as const, if it is constexpr

Comment (sunverwerth) [reply Userland/DevTools/Profiler/SourceModel.cpp]: Done!

File: Userland/DevTools/Profiler/Profile.cpp [hunk 2]
```diff
@@ -554,6 +555,23 @@ GUI::Model* Profile::disassembly_model()
     return m_disassembly_model;
 }
 
+void Profile::set_source_index(const GUI::ModelIndex& index)
```
Comment (linusg) [inline Userland/DevTools/Profiler/Profile.cpp:558]: Please use east const (everywhere below as well)

File: Userland/DevTools/Profiler/SourceModel.cpp [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+#include "SourceModel.h"
+#include "Profile.h"
+#include <LibDebug/DebugInfo.h>
+#include <LibGUI/Painter.h>
+#include <LibSymbolication/Symbolication.h>
+#include <stdio.h>
+
+namespace Profiler {
+
+static constexpr StringView c_source_root_path = "/usr/src/serenity/"sv;
```
Comment (linusg) [inline Userland/DevTools/Profiler/SourceModel.cpp:16]: `c_`?

File: Userland/DevTools/Profiler/SourceModel.cpp [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+        if (maybe_file.is_error()) {
+            dbgln("Could not map source file \"{}\". Tried {}. {} (errno={})", filename, source_file_name, maybe_file.error().string_literal(), maybe_file.error().code());
+            return;
+        }
+
+        auto file = maybe_file.value();
+
+        while (!file->eof()) {
+            m_lines.append({ file->read_line(1024), 0 });
+        }
```
Comment (linusg) [inline Userland/DevTools/Profiler/SourceModel.cpp:40]: redundant curly braces

```suggestion
        while (!file->eof())
            m_lines.append({ file->read_line(1024), 0 });
```

File: Userland/DevTools/Profiler/SourceModel.cpp [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+    static RefPtr<Gfx::Bitmap> bitmap;
+    if (!bitmap) {
+        bitmap = Gfx::Bitmap::try_create(Gfx::BitmapFormat::BGRx8888, { 101, 1 }).release_value_but_fixme_should_propagate_errors();
+        GUI::Painter painter(*bitmap);
+        painter.fill_rect_with_gradient(Orientation::Horizontal, bitmap->rect(), Color::from_rgb(0xffc080), Color::from_rgb(0xff3000));
+    }
+    return *bitmap;
+}
+
+static Color color_for_percent(int percent)
```
Comment (linusg) [inline Userland/DevTools/Profiler/SourceModel.cpp:68]: should probably be be an u8?

File: Userland/DevTools/Profiler/SourceModel.cpp [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+            return line.location;
+        }
+
+        if (index.column() == Column::LineNumber) {
+            return line.line_number;
+        }
+
+        if (index.column() == Column::SourceCode) {
+            return line.source_code;
+        }
```
Comment (linusg) [inline Userland/DevTools/Profiler/SourceModel.cpp:214]: No curly braces here

**Pull request merged**